lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+use std::ffi::OsStr;
+use std::path::Path;
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use phylum_types::types::package::PackageType;
+use purl::GenericPurl;
+use serde::Deserialize;
+use thiserror::Error;
+
+use crate::{Package, PackageVersion, Parse, ThirdPartyVersion};
+
+// Define a custom error for unknown ecosystems.
+#[derive(Error, Debug)]
+#[error("Could not determine ecosystem")]
+struct UnknownEcosystem;
+
+// Define the generic trait for components.
+trait Component {
+    fn component_type(&self) -> &str;
+    fn name(&self) -> &str;
+    fn version(&self) -> &str;
+    fn scope(&self) -> Option<&str>;
+    fn purl(&self) -> Option<&str>;
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized;
+}
+
+// CycloneDX BOM
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Bom<T> {
+    components: Option<T>,
+}
+
+// Struct for wrapping a list of components from XML
+#[derive(Clone, Debug, Deserialize)]
+struct Components<T> {
+    #[serde(rename = "component")]
+    components: Vec<T>,
+}
+
+// Represents a single XML component.
+#[derive(Clone, Debug, Deserialize)]
+struct XmlComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Components<XmlComponent>>,
+}
+
+impl Component for XmlComponent {
+    fn component_type(&self) -> &str {
+        &self.component_type
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn version(&self) -> &str {
+        &self.version
+    }
+
+    fn scope(&self) -> Option<&str> {
+        self.scope.as_deref()
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_ref().map(|comps| comps.components.as_slice())
+    }
+}
+
+// Represents a single JSON component.
+#[derive(Clone, Debug, Deserialize)]
+struct JsonComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Vec<JsonComponent>>,
```

Comment:
Why is this an option? Can't we just default to empty vec? We're unwrapping to an empty vec later anyway.